### PR TITLE
Add support for self tests.

### PR DESCRIPTION
**Pull request by nomis52**

### CODE REVIEW

File: firmware/src/dimmer_model.c [hunk 1]
```diff
@@ -241,7 +263,21 @@ uint8_t *AddStatusMessageToResponse(uint8_t *ptr,
...
+  message->is_active = true;
+  message->sub_device = sub_device_index;
+  message->status_type = status_type;
+  message->message_id = status_id;
+  message->data_value1 = data_value1;
+  message->data_value2 = data_value2;
+}
+
+void QueueSubDeviceStatusMessage(DimmerSubDevice *device,
                         RDMStatusType status_type,
```
Comment (peternewman) [inline firmware/src/dimmer_model.c]: Fix the spacing.

File: firmware/src/dimmer_model.c [hunk 2]
```diff
@@ -313,13 +357,60 @@ int DimmerModel_GetStatusMessages(const RDMHeader *header,
...
-                            STS_TESTING_DESCRIPTION,
+                            STS_OLP_TESTING_DESCRIPTION,
+                            RDM_DEFAULT_STRING_SIZE);
+  return RDMResponder_AddHeaderAndChecksum(header, ACK, ptr - g_rdm_buffer);
+}
+
+int DimmerModel_GetSelfTest(const RDMHeader *header,
+                            UNUSED const uint8_t *param_data) {
+  uint8_t *ptr = g_rdm_buffer + sizeof(RDMHeader);
+  *ptr++ = g_root_device.running_self_test ? true : false;
```
Comment (peternewman) [inline firmware/src/dimmer_model.c]: Does this need casting?

File: firmware/src/dimmer_model.c [hunk 3]
```diff
@@ -313,13 +357,60 @@ int DimmerModel_GetStatusMessages(const RDMHeader *header,
...
+    g_root_device.running_self_test = self_test_id;
+    g_root_device.self_test_timer = CoarseTimer_GetTime();
+  }
+  return RDMResponder_BuildSetAck(header);
+}
+
+int DimmerModel_GetSelfTestDescription(const RDMHeader *header,
+                                       UNUSED const uint8_t *param_data) {
+  const uint8_t self_test_id = param_data[0];
+  if (self_test_id == 0 || self_test_id > NUMBER_OF_SELF_TESTS) {
```
Comment (peternewman) [inline firmware/src/dimmer_model.c]: == SELF_TEST_OFF

File: firmware/src/rdm.h [hunk 4]
```diff
@@ -816,6 +821,12 @@ typedef enum {
   STS_DMXNSC_OK = 0x0083,
 } RDMStatusMessageId;
 
+typedef enum {
```
Comment (peternewman) [inline firmware/src/rdm.h]: Add these to https://wiki.openlighting.org/index.php/Open_Lighting_Allocations

File: firmware/src/dimmer_model.c (hunk 1 above)
Comment (nomis52) [inline firmware/src/dimmer_model.c]: Done

File: firmware/src/dimmer_model.c (hunk 2 above)
Comment (nomis52) [inline firmware/src/dimmer_model.c]: No, I don't think so.

File: firmware/src/dimmer_model.c (hunk 3 above)
Comment (nomis52) [inline firmware/src/dimmer_model.c]: Done

File: firmware/src/rdm.h (hunk 4 above)
Comment (nomis52) [inline firmware/src/rdm.h]: Done

File: firmware/src/dimmer_model.c (hunk 2 above)
Comment (peternewman) [inline firmware/src/dimmer_model.c]: Okay

**Comment by peternewman** [comment]
LGTM

**Pull request merged**